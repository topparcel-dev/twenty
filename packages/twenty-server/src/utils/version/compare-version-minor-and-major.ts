import * as semver from 'semver';

export function compareVersionMajorAndMinor(
  rawVersion1: string,
  rawVersion2: string,
): -1 | 0 | 1 {
  const [version1, version2] = [rawVersion1, rawVersion2].map((version) =>
    semver.parse(version),
  );

  if (version1 === null || version2 === null) {
    throw new Error(`Received invalid version: ${rawVersion1} ${rawVersion2}`);
  }

  const v1WithoutPatch = `${version1.major}.${version1.minor}.0`;
  const v2WithoutPatch = `${version2.major}.${version2.minor}.0`;

  return semver.compare(v1WithoutPatch, v2WithoutPatch);
}
