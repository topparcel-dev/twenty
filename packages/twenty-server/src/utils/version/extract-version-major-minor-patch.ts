import semver from "semver";

export const extractVersionMajorMinorPatch = (version: string) => {
    const parsed = semver.parse(version);

    if (parsed === null) {
        return null
    }

    return `${parsed.major}.${parsed.minor}.${parsed.patch}`;
}