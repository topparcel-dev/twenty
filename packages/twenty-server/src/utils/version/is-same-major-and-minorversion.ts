import * as semver from 'semver';

export function isSameMajorAndMinorVersion(rawVersion1: string, rawVersion2: string): boolean {
  const [version1, version2] = [rawVersion1, rawVersion2].map(
    (version) => semver.parse(version),
  );

  if (version1 === null || version2 === null) {
    throw new Error(`Received invalid version: ${rawVersion1} ${rawVersion2}`);
  }

  const diffMajor = version1.major !== version2.major;
  const diffMinor = version1.minor !== version2.minor;
  return !diffMajor && !diffMinor
}
