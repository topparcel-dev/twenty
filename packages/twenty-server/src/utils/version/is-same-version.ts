import * as semver from 'semver';

export function isSameVersion(version1: string, version2: string): boolean {
  const invalidVersion = [version1, version2].filter(
    (version) => semver.parse(version) === null,
  );
  console.log(invalidVersion);
  if (invalidVersion.length > 0) {
    throw new Error(`Received invalid version: ${invalidVersion.join(' ')}`);
  }

  const result = semver.compare(version1, version2);

  return result === 0;
}
