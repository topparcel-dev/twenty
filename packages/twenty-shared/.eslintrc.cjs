module.exports = {
  extends: ['../../.eslintrc.cjs'],
  ignorePatterns: ['!**/*'],
  overrides: [
    {
      files: ['**/*.ts'],
      parserOptions: {
        project: ['packages/twenty-shared/tsconfig.*.json'],
      },
      rules: {
        '@nx/dependency-checks': 'error',
      },
    },
  ],
};
