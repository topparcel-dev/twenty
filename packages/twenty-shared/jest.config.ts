import { JestConfigWithTsJest, pathsToModuleNameMapper } from 'ts-jest';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const tsConfig = require('./tsconfig.json');

const jestConfig: JestConfigWithTsJest = {
  displayName: 'twenty-ui',
  preset: '../../jest.preset.js',
  testEnvironment: 'jsdom',
  transformIgnorePatterns: ['../../node_modules/'],
  transform: {
    '^.+\\.[tj]sx?$': [
      '@swc/jest',
      {
        jsc: {
          parser: { syntax: 'typescript', tsx: true },
          transform: { react: { runtime: 'automatic' } },
        },
      },
    ],
  },
  moduleNameMapper: {
    '\\.(jpg|jpeg|png|gif|webp|svg|svg\\?react)$':
      '<rootDir>/__mocks__/imageMock.js',
    ...pathsToModuleNameMapper(tsConfig.compilerOptions.paths),
  },
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx'],
  extensionsToTreatAsEsm: ['.ts', '.tsx'],
  coverageDirectory: './coverage',
  coverageThreshold: {
    global: {
      statements: 100,
      lines: 100,
      functions: 100,
    },
  },
};

export default jestConfig;
