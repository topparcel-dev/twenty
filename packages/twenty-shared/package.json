{
  "name": "twenty-shared",
  "version": "0.42.0-canary",
  "license": "AGPL-3.0",
  "main": "./dist/index.js",
  "scripts": {
    "build": "npx vite build"
  },
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "import": "./dist/index.mjs",
      "require": "./dist/index.js"
    }
  },
  "engines": {
    "node": "^18.17.1",
    "npm": "please-use-yarn",
    "yarn": "^4.0.2"
  }
}
