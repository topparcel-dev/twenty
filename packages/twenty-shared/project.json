{
  "name": "twenty-shared",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/twenty-shared/src",
  "projectType": "library",
  "tags": ["scope:shared"],
  "targets": {
    "build": {
      "dependsOn": ["generateBarrels", "^build"],
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "{projectRoot}/dist"
      }
    },
    "generateBarrels": {
      "executor": "nx:run-commands",
      "cache": true,
      "inputs": ["production", "{projectRoot}/scripts/generateBarrels.ts"],
      "outputs": [
        "{projectRoot}/src/index.ts",
        "{projectRoot}/src/*/index.ts",
        "{projectRoot}/package.json"
      ],
      "options": {
        "command": "tsx {projectRoot}/scripts/generateBarrels.ts"
      }
    },
    "typecheck": {},
    "test": {},
    "lint": {
      "options": {
        "lintFilePatterns": [
          "{projectRoot}/src/**/*.{ts,tsx,json}",
          "{projectRoot}/package.json"
        ],
        "reportUnusedDisableDirectives": "error"
      },
      "configurations": {
        "fix": {}
      }
    },
    "fmt": {
      "options": {
        "files": "src"
      },
      "configurations": {
        "fix": {}
      }
    }
  }
}
