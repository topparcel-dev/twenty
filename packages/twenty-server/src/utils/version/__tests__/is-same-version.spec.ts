import { EachTestingContext } from 'twenty-shared';

import { isSameVersion } from 'src/utils/version/is-same-version';

type IsSameVersionTestCase = EachTestingContext<{
  version1: string;
  version2: string;
  expected?: boolean;
  expectToThrow?: boolean;
}>;
describe('is-same-version', () => {
  const differentVersionTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: '1.0.0',
        version2: '1.1.0',
      },
      title: 'different minor version',
    },
    {
      context: {
        version1: '2.3.0',
        version2: '2.4.0',
      },
      title: 'different minor version',
    },
    {
      context: {
        version1: '0.1.0',
        version2: '0.2.0',
      },
      title: 'different minor version with zero major',
    },
    {
      context: {
        version1: '2.3.5',
        version2: '2.4.1',
      },
      title: 'different minor and patch versions',
    },
    {
      context: {
        version1: '1.0.0-alpha',
        version2: '1.1.0-beta',
      },
      title: 'different minor version with different pre-release tags',
    },
    {
      context: {
        version1: 'v1.0.0',
        version2: 'v1.1.0',
      },
      title: 'different minor version with v prefix',
    },
  ];

  const sameVersionTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: '1.1.0',
        version2: '1.1.0',
        expected: true,
      },
      title: 'exact same version',
    },
    {
      context: {
        version1: 'v1.1.0',
        version2: 'v1.1.0',
        expected: true,
      },
      title: 'exact same version with v prefix',
    },
    {
      context: {
        version1: '1.1.0-alpha',
        version2: '1.1.0-alpha',
        expected: true,
      },
      title: 'exact same version with same pre-release tag',
    },
    {
      context: {
        version1: '0.0.1',
        version2: '0.0.1',
        expected: true,
      },
      title: 'exact same version with all zeros',
    },
    {
      context: {
        version1: 'v1.1.0',
        version2: '1.1.0',
        expected: true,
      },
      title: 'same version with different v-prefix',
    },
  ];

  const invalidTestCases: IsSameVersionTestCase[] = [
    {
      context: {
        version1: 'invalid',
        version2: '1.1.0',
        expectToThrow: true,
      },
      title: 'invalid version1',
    },
    {
      context: {
        version1: '1.0.0',
        version2: 'invalid',
        expectToThrow: true,
      },
      title: 'invalid version2',
    },
    {
      context: {
        version1: '1.0',
        version2: '1.1.0',
        expectToThrow: true,
      },
      title: 'incomplete version1',
    },
  ];

  test.each([
    ...sameVersionTestCases,
    ...invalidTestCases,
    ...differentVersionTestCases,
  ])(
    '$title',
    ({
      context: { version1, version2, expected = false, expectToThrow = false },
    }) => {
      if (expectToThrow) {
        expect(() =>
          isSameVersion(version1, version2),
        ).toThrowErrorMatchingSnapshot();
      } else {
        expect(isSameVersion(version1, version2)).toBe(expected);
      }
    },
  );
});
