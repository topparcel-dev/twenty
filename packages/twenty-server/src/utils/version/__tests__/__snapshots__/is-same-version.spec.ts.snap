// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`is-same-version incomplete version1 1`] = `"Received invalid version: 1.0"`;

exports[`is-same-version invalid version1 1`] = `"Received invalid version: invalid"`;

exports[`is-same-version invalid version2 1`] = `"Received invalid version: invalid"`;
